ddl/table.go

```diff
@@ -152,6 +138,138 @@ func onCreateTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 	}
 }
 
+func inheritPlacementRuleFromDB(tbInfo *model.TableInfo, dbInfo *model.DBInfo) error {
+	// Can not use both a placement policy and direct assignment. If you alter specify both in a CREATE TABLE or ALTER TABLE an error will be returned.
+	if tbInfo.DirectPlacementOpts != nil && tbInfo.PlacementPolicyRef != nil {
+		return ErrPlacementPolicyWithDirectOption.GenWithStackByArgs(tbInfo.PlacementPolicyRef.Name)
+	}
+	if tbInfo.DirectPlacementOpts == nil && tbInfo.PlacementPolicyRef == nil {
+		if dbInfo.DirectPlacementOpts != nil {
+			clone := *dbInfo.DirectPlacementOpts
+			tbInfo.DirectPlacementOpts = &clone
+		}
+		if dbInfo.PlacementPolicyRef != nil {
+			clone := *dbInfo.PlacementPolicyRef
+			tbInfo.PlacementPolicyRef = &clone
+		}
+	}
+	return nil
+}
+
+func getBundleFromTblInfo(t *meta.Meta, job *model.Job, tbInfo *model.TableInfo) (*placement.Bundle, error) {
+	tableBundleWrapper := func(bundle *placement.Bundle) *placement.Bundle {
+		bundle.Index = placement.RuleIndexTable
+		bundle.Override = true
+		return bundle
+	}
+	var (
+		bundle *placement.Bundle
+		err    error
+	)
+	if tbInfo.DirectPlacementOpts != nil {
+		// check the direct placement option compatibility.
+		if err = checkPolicyValidation(tbInfo.DirectPlacementOpts); err != nil {
+			job.State = model.JobStateCancelled
+			return nil, errors.Trace(err)
+		}
+		// build bundle from direct placement options.
+		bundle, err = placement.NewBundleFromOptions(tbInfo.DirectPlacementOpts)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return nil, errors.Trace(err)
+		}
+	}
+	if tbInfo.PlacementPolicyRef != nil {
+		// placement policy reference will override the direct placement options.
+		po, err := checkPlacementPolicyExistAndCancelNonExistJob(t, job, tbInfo.PlacementPolicyRef.ID)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return nil, errors.Trace(infoschema.ErrPlacementPolicyNotExists.GenWithStackByArgs(tbInfo.PlacementPolicyRef.Name))
+		}
+		// build bundle from placement policy.
+		bundle, err = placement.NewBundleFromOptions(po.PlacementSettings)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return nil, errors.Trace(err)
+		}
+	}
+	if bundle == nil {
+		return nil, nil
+	}
+	err = bundle.Tidy()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```

Comment:
`Reset` first, then `Tidy`. Maybe we could move all `Tidy` to `infosync.PutRuleBundles`.

Reply:
Tidy first is reasonable. Since tidy wanna append a `NotIn TiFlash` rule, it should be done before reset work.
The later will clone all these rule for other new ids. otherwise, we should append `NotIn TiFlash` rule for every independent id inferred from startKey/ EndKey

Reply:
Tidy was intended to be used as an post-optimization. Maybe we should move the append of `-engine=tiflash` into reset.

Reply:
I put it into the new bundle portal function --- `NewBundleFromOptions`, which are much more clear.